analytic_engine/src/meta/details.rs

```diff
@@ -291,86 +327,231 @@ impl Manifest for ManifestImpl {
         &self,
         request: MetaUpdateRequest,
     ) -> std::result::Result<(), Box<dyn std::error::Error + Send + Sync>> {
-        let location = request.location;
-        self.store_update_to_wal(request).await?;
-
-        let num_updates = self
-            .num_updates_since_snapshot
-            .fetch_add(1, Ordering::Relaxed);
-        if num_updates < self.opts.snapshot_every_n_updates {
-            return Ok(());
-        }
+        debug!("try to store update:{:?}", request);
 
-        self.maybe_do_snapshot(location).await?;
+        let table_id = request.meta_update.table_id();
+        let location = WalLocation::new(
+            // TODO: use shard id as region id.
+            table_id.as_u64(),
+            request.shard_info.cluster_version,
+            table_id.as_u64(),
+        );
+        let space_id = request.meta_update.space_id();
+        let table_id = request.meta_update.table_id();
+        self.store_update_to_wal(request.meta_update, location)
+            .await?;
+
+        self.maybe_do_snapshot(space_id, table_id, location, false)
+            .await?;
 
         Ok(())
     }
 
     async fn load_data(
         &self,
-        location: WalLocation,
-        do_snapshot: bool,
+        load_req: &LoadRequest,
     ) -> std::result::Result<Option<TableManifestData>, Box<dyn std::error::Error + Send + Sync>>
     {
-        if do_snapshot {
-            if let Some(snapshot) = self.maybe_do_snapshot(location).await? {
+        let location = WalLocation::new(
+            // TODO: use shard id as region id.
+            load_req.table_id.as_u64(),
+            load_req.cluster_version,
+            load_req.table_id.as_u64(),
+        );
+
+        if load_req.do_snapshot {
+            if let Some(snapshot) = self
+                .maybe_do_snapshot(load_req.space_id, load_req.table_id, location, true)
+                .await?
+            {
                 return Ok(snapshot.data);
             }
         }
```

Comment:
Maybe, we remove it?
I think `load_data` should be a read-only operation?

Reply:
done